### PR TITLE
add solveCliqDown!

### LINKED ISSUE #1170: solveCliq feature update
Just documenting what we spoke about...
- [x] record cliques replace old solveCliq function
- [x] solve clique will do up/down
- [x] in messages can be created from fg or input directly
- [x] out messages will be output
- [x] EDIT, DF, option to bring out CSM history vector for step by step analysis.
- therefore, it becomes more of a debug feature on how a clique is solved with different messages
- we can add extra plumbing later to ease clique interaction.

### PR DESCRIPTION
**Pull request by Affie**
close #1170

### CODE REVIEW

File: src/SolverAPI.jl [hunk 1]
```diff
@@ -558,3 +563,79 @@ function solveCliqUp!(fg::AbstractDFG,
 
   return hist, beliefMessageOut
 end
+
+"""
+Internal function used for solveCliqDown!
```
Comment (dehann) [inline src/SolverAPI.jl:568]: Perhaps just specify Tx or Rx at this `cliq` vs `parent`

File: src/SolverAPI.jl [hunk 2]
```diff
@@ -558,3 +563,79 @@ function solveCliqUp!(fg::AbstractDFG,
...
+"""
+function _buildMessageDown(fg, tree, cliqid)
+  cliq = getClique(tree, cliqid)
+  parent = getParent(tree, cliq)[1]
+  return getCliqDownMsgsAfterDownSolve(fg, parent; status=DOWNSOLVED)
```
Comment (dehann) [inline src/SolverAPI.jl:573]: what if the clique has status `MARGINALIZED`, then `DOWNSOLVED` is not entirely correct.  Is there perhaps an easy way to use the actual clique status vs hardcoded?

Comment (dehann) [reply src/SolverAPI.jl:573]: or via keyword to allow user to override the default status?

File: src/SolverAPI.jl [hunk 3]
```diff
@@ -558,3 +563,79 @@ function solveCliqUp!(fg::AbstractDFG,
...
+      messages[e.dst] = takeBeliefMessageDown!(tree, e) 
+    end
+    messages
+  end
+
+  recordcliqs = recordcliq ?  [getFrontals(cliq)[1]] : Symbol[]
+
+  hist = tryCliqStateMachineSolve!(fg, tree, cliq.id; 
+                                   verbose=verbose, drawtree=opt.drawtree, limititers=opt.limititers,
+                                   recordcliqs=recordcliqs, incremental=opt.incremental)
```
Comment (dehann) [inline src/SolverAPI.jl:632]: in future versions we can bring back the `async` option.  Just keeping a note here.

File: src/TreeMessageUtils.jl [hunk 4]
```diff
@@ -574,7 +574,7 @@ during clique inference.
 
 DevNotes
 - TODO make `::Vector{Symbol}` version.
-
+- TODO function taking fcts::Vector{DFGFactor} is unused and replace by the tags version, perhaps we can remove it.
```
Comment (dehann) [inline src/TreeMessageUtils.jl:577]: just making sure, `tags` or `labels`?

File: src/SolverAPI.jl (hunk 2 above)
Comment (dehann) [reply src/SolverAPI.jl:573]: also e.g.:
https://github.com/JuliaRobotics/IncrementalInference.jl/pull/1186/files#diff-bdbb86421fa7a813967257eb8845a821e8595683ead3f4615750646db435d942R640

File: src/TreeMessageUtils.jl (hunk 4 above)
Comment (Affie) [reply src/TreeMessageUtils.jl:577]: tags, If you remember we remove messages based on the `__LIKELIHOOD_MESSAGE__` tag

**Review by dehann — APPROVED** [review]
Looks good thanks.  Perhaps just take a look at the comments and see if any are useful or valid.

**Comment by dehann** [comment]
CI on 1.5 failed due to numerical error `(1.15 < 1.0 )`.  I restarted it.  The new multihypo updates (although not included here yet) might fail on old obsolete tests, and there are too many to search so I'm checking all CI outcomes instead.

File: src/SolverAPI.jl (hunk 2 above)
Comment (Affie) [reply src/SolverAPI.jl:573]: There is no parent clique, so no easy way. 
We can do kwarg, but does it matter? Downsolve is treated the same in all cases.

Comment (dehann) [reply src/SolverAPI.jl:573]: i'd suggest just going with kwarg to make the hard coded a little better.  We'll see if tree CSM logic evolves to a point where DOWNSOLVE vs MARGINALIZED or others make a difference -- agreed not 100% yet today, but I have reason to suspect things will change there over time.

File: src/SolverAPI.jl (hunk 1 above)
Comment (Affie) [reply src/SolverAPI.jl:568]: I'm not following this comment, do you want clearer comments on whether it is an out or in message?

Comment (Affie) [reply src/SolverAPI.jl:568]: put is Tx and take! is Rx

File: src/SolverAPI.jl [hunk 5]
```diff
@@ -410,73 +410,16 @@ See `solveTree!`.
...
-printCliqHistorySummary(hist)
-```
-
-DevNotes
-- on `sandboxStateMachineStep`, see FSM #42
-
-Related
-
-[`solveTree!`](@ref), [`buildTreeReset!`](@ref), [`printCliqHistorySummary`](@ref), [`repeatCSMStep!`](@ref), `sandboxStateMachineStep`
+Internal function used for solveCliqUp! to build the incomming upward message (Rx)
```
Comment (dehann) [inline src/SolverAPI.jl:416]:
```suggestion
Internal function used for solveCliqUp! to build the incoming upward message (Rx)
```

File: src/SolverAPI.jl [hunk 6]
```diff
@@ -558,3 +504,78 @@ function solveCliqUp!(fg::AbstractDFG,
 
   return hist, beliefMessageOut
 end
+
+"""
+    $SIGNATURES
+Internal function used for solveCliqDown! to build the incomming downward message (Rx)
```
Comment (dehann) [inline src/SolverAPI.jl:510]:
```suggestion
Internal function used for solveCliqDown! to build the incoming downward message (Rx)
```

Comment (Affie) [reply src/SolverAPI.jl:510]: Thanks, I need a spell checker...

**Comment by Affie** [comment]
Test passed, just spelling from code review. So not waiting for the new test.

**Pull request merged**